internal/pkg/vm/internal/runtime/runtime.go

```diff
@@ -8,27 +8,163 @@ import (
 	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/address"
 )
 
+
+
+// Runtime2 is the VM.
+// Review: the object that implements this outlives actor calls.
+type Runtime2 struct {
+	CurerntEpoch() Epoch
+	// Review: why is this taking an offset? verify size of offset
+	Randomness(epoch Epoch, offset uint64) Randomness
```

Comment:
Yes, that's the intent -- we don't want accidental collisions when different computations (or different phases during a single computation) need to draw randomness from the same epoch. Ideally, the offsets would be aggregated in a central place to make it easier to enforce uniqueness, similar to `MethodNum`.

Reply:
Got it.

Reply:
If that case is the case I consider having a `type RandomnessEvent uint64`, and change the signature to `Randomness(epoch Epoch, event RandomnessEvent) Randomness`.

That being said, those offsets are not all hardcoded. The `InitActor` seems to call `Randomess/2` with the `ActorID` as an argument. 

Bonus question: Are the actors allowed to call `Randomness/2` with a first argument other than `Current Epoch`?

Reply:
> Bonus question: Are the actors allowed to call Randomness/2 with a first argument other than Current Epoch?

Reply:
> Why not just let the caller do the arithmetic of epoch - offset and reduce to a single parameter?

The offset here isn't measured in epochs; rather, each epoch `E` determines a pseudorandom stream, and then the user may provide an offset into that stream.

Reply:
> consider having a type RandomnessEvent uint64, and change the signature to Randomness(epoch Epoch, event RandomnessEvent) Randomness

I think this would be a good approach, if the event is the _only_ seed we need. However, it sounds like that's not the case (e.g., if `ActorID` is used as an argument). Given that requirement, I'd propose something like the following:

```
type RandomSeed []byte
type Serialization []byte
Randomness(epoch Epoch, seed RandomSeed)
DeriveRandomSeedExtended(event RandomnessEvent, s Serialization)
DeriveRandomSeed(event RandomnessEvent)
Serialize_T(T) Serialization  // for each type T, may be automatically generated in the future
```

What do you think?

Reply:
Not sure why we would need serialization. 
I thought you wanted an offset into the random stream, why are you redefining it as a `[]byte`?

```go
type seed interface {
    seed() uint64
}

func Randomness(epoch Epoch, seeder seed) {}

func (x RandomnessEvent) seed() uint64 {
    // impl...
}

func (x ActorID) seed() uint64 {
    // impl...
}
```

Reply:
> Not sure why we would need serialization.

This is to avoid errors that might arise from duplication of code otherwise. E.g., if we have:

```go
type Foo struct {
  A int
  B int
}
```

and we would like to use objects of type `Foo` directly as random seeds, then in the underlying implementation, we want to derive some integer seed from any `Foo` instance such that distinct instances produce distinct seeds. `Serialize` (followed by a collision-resistant hash) and `CID` would both satisfy this criterion, so we shouldn't add a third independent implementation for it as well. (In fact, ideally, CID should also be implemented in terms of serialization.)

Reply:
> I thought you wanted an offset into the random stream, why are you redefining it as a []byte?

I think this is just a question of notation -- if we model it as an _infinite_ stream, any bit string can be used as an integer to index into it. But for practical purposes, we can model it as a finite stream of length `2^64`, and hash the `[]byte` index into that range instead.